File: ruby-api/app/controllers/devs_controller.rb
class DevsController < ApplicationController
  def index
    devs = Dev.all

    render json: devs
  end

  def show
    dev = Dev.where(username: params[:username].upcase).first

    render json: dev
  end

  def create
    dev = Dev.new(dev_params)

    Dev.upsert(dev.attributes, unique_by: :username)
    render json: { message: "Notes saved!", username: dev.username, notes: dev.notes }
  end

  def dev_params
    dev_params = params.require(:dev).permit(:username, :notes)
    dev_params = { username: dev_params[:username].upcase, notes: dev_params[:notes] }
  end
end

File: node-api/src/controller/DevsController.ts
import { Request, Response } from "express";

import connection from "../database/connection";
import { DevProps } from "../interfaces";

interface CustomRequest extends Request {
  body: DevProps;
}

export default {
  async index(request: Request, response: Response): Promise<Response> {
    const dev = await connection<DevProps>("devs").select("*");

    return response.json(dev);
  },

  async show(request: Request, response: Response): Promise<Response> {
    const { username } = request.params;

    const dev = await connection<DevProps>("devs")
      .where("username", "=", String(username).toUpperCase())
      .select("*")
      .first();

    return response.json(dev);
  },

  async save(request: CustomRequest, response: Response): Promise<Response> {
    const { username, notes } = request.body;

    await connection<DevProps>("devs")
      .insert({
        username: username.toUpperCase(),
        notes,
      })
      .onConflict("username")
      .merge();

    return response.json({ message: "Notes saved!", username, notes });
  },
};
